### ISSUE TITLE
Python logging observer fails to log Failures

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8316]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)| @mithrandi reported|
|-|-|
|Trac ID|trac#8316|
|Type|defect|
|Created|2016-05-03 01:48:25Z|
|Branch|https://github.com/twisted/twisted/tree/stdlib-exception-logging-8316|

```
import logging
from twisted.python.log import startLoggingWithObserver, PythonLoggingObserver
from twisted.python.failure import Failure
from twisted.logger import Logger

logging.basicConfig(level=logging.DEBUG)
startLoggingWithObserver(PythonLoggingObserver().emit, False)
log = Logger()

try:
    1 / 0
except Exception:
    log.failure('hi mom', Failure())
```

produces:

```
$ python logging_test.py 
INFO:twisted:Log opened.
CRITICAL:twisted:hi mom
```

By comparison, the following code:
```
import sys
from twisted.python.log import startLogging
from twisted.python.failure import Failure
from twisted.logger import Logger

startLogging(sys.stdout)
log = Logger()

try:
    1 / 0
except Exception:
    log.failure('hi mom', Failure())
```
produces:
```
> python logging_test.py 
2016-05-03 03:47:16+0200 [-] Log opened.
2016-05-03 03:47:16+0200 [-] hi mom
        Traceback (most recent call last):
        --- <exception caught here> ---
          File "logging_test.py", line 10, in <module>
            1 / 0
        exceptions.ZeroDivisionError: integer division or modulo by zero
```

<details><summary>Searchable metadata</summary>

```
trac-id__8316 8316
type__defect defect
reporter__mithrandi mithrandi
priority__normal normal
milestone__None None
branch__branches_stdlib_exception_logging_8316 branches/stdlib-exception-logging-8316
branch_author__mithrandi mithrandi
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1462240105330869 1462240105330869
changetime__1462604534459113 1462604534459113
version__None None
owner__mithrandi mithrandi

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_1"></a>|@mithrandi commented|
|-|-|

(In [47346]) Branching to stdlib-exception-logging-8316.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_2"></a>|@mithrandi commented|
|-|-|

I sort of expected this not to work with cleaned failures, but it turns out it Just Works™ (there's a test and everything). Most of the builders have checked in green at the time of writing this.

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @mithrandi|
|-|-|

Looks good. Thanks!

Just a few minor comments


1. I don't understand why the test_cleanedFailure is related to twisted/logger/_stdlib.py ? and what was changed from this patch is affecting the behaviour of Failure.cleanFailure()

------

2. It would be nice to include a broader assertion for the output... for example `Hi mom Traceback (most recent call last):` is part of the output

------

Please check my comments and resubmit for review.

I prefer not to have unrelated tests as this might confuse developers during a future refactoring.

It can also mean that I don't fully understand your patch :) 


Thanks again for your contribution!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_4"></a>|@mithrandi commented|
|-|-|

`Failure.cleanFailure` replaces the real traceback/frames on the `Failure` with some fake objects that look like a real traceback but don't have references to the frame objects in the traceback and thus don't keep all of the local variables alive. The test is just to make sure that stdlib logging can handle getting one of those fake tracebacks instead of a real traceback object. I think the test docstring isn't very clear about this, so I'll expand it.

I'll expand the output assertions a bit more, too.

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)|@mithrandi removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_6"></a>|@adiroiban set owner to @mithrandi|
|-|-|

please merge. thanks!

**Comment by twisted-trac** [comment]
|[<img alt="mithrandi's avatar" src="https://avatars.githubusercontent.com/u/856430?s=50" width="50" height="50">](https://github.com/mithrandi)<a name="note_7"></a>|@mithrandi set status to `closed`|
|-|-|

(In [47357]) Merge stdlib-exception-logging-8316: Fix logging of exceptions to stdlib logging.

Author: mithrandi
Reviewer: adiroiban
Fixes: [#8316](https://github.com/twisted/twisted/issues/8316)

Exceptions weren't logged at all; now they are.
